### ISSUE TITLE
OOM may occur in red5 when presenter does rapid starting/stopping of desktop sharing 

### ISSUE THREAD

**Issue by ffdixon** [bigbluebutton/bigbluebutton#3047]
We've seen some cases when the presenter is starting/stopping desktop sharing,  red5 (or more specifically mina) will not flush the outgoing buffers quickly enough and may cause the JVM can run out of heap space.

The circumstances seem to be there are (a) a few dozen users in the session, (b) some users are on low bandwidth connections (this causes the buffers to grow larger), and (c) the presenter starts/stop their desktop sharing frequently.

We've been able to reproduce with (a), (b), and (c)-- though it takes quite a few start/stop in a session to the OOM to trigger.

Here's a sample exception from red5's error.log when this occurs:

```
2016-03-05 21:58:12,086Z [ForkJoinPool-1-worker-4] ERROR o.r.s.m.InMemoryPushPushPipe - Exception pushing message to consumer
java.lang.OutOfMemoryError: Java heap space
        at java.nio.HeapByteBuffer.<init>(HeapByteBuffer.java:57) ~[na:1.7.0_95]
        at java.nio.ByteBuffer.allocate(ByteBuffer.java:331) ~[na:1.7.0_95]
        at org.apache.mina.core.buffer.SimpleBufferAllocator.allocateNioBuffer(SimpleBufferAllocator.java:42) ~[mina-core-2.0.9.jar:na]
        at org.apache.mina.core.buffer.SimpleBufferAllocator.allocate(SimpleBufferAllocator.java:34) ~[mina-core-2.0.9.jar:na]
        at org.apache.mina.core.buffer.IoBuffer.allocate(IoBuffer.java:227) ~[mina-core-2.0.9.jar:na]
        at org.red5.server.net.rtmp.codec.RTMPProtocolEncoder.encodePacket(RTMPProtocolEncoder.java:157) ~[red5-server-common-1.0.6-SNAPSHOT.jar:1.0.6-SNAPSHOT]
        at org.red5.server.net.rtmp.codec.RTMPProtocolEncoder.encode(RTMPProtocolEncoder.java:109) ~[red5-server-common-1.0.6-SNAPSHOT.jar:1.0.6-SNAPSHOT]
        at org.red5.server.net.rtmp.codec.RTMPMinaProtocolEncoder.encode(RTMPMinaProtocolEncoder.java:69) ~[red5-server-common-1.0.6-SNAPSHOT.jar:1.0.6-SNAPSHOT]
        at org.apache.mina.filter.codec.ProtocolCodecFilter.filterWrite(ProtocolCodecFilter.java:306) ~[mina-core-2.0.9.jar:na]
        at org.apache.mina.core.filterchain.DefaultIoFilterChain.callPreviousFilterWrite(DefaultIoFilterChain.java:625) ~[mina-core-2.0.9.jar:na]
        at org.apache.mina.core.filterchain.DefaultIoFilterChain.access$1500(DefaultIoFilterChain.java:48) ~[mina-core-2.0.9.jar:na]
        at org.apache.mina.core.filterchain.DefaultIoFilterChain$EntryImpl$1.filterWrite(DefaultIoFilterChain.java:953) ~[mina-core-2.0.9.jar:na]
        at org.apache.mina.core.filterchain.DefaultIoFilterChain$TailFilter.filterWrite(DefaultIoFilterChain.java:877) ~[mina-core-2.0.9.jar:na]
        at org.apache.mina.core.filterchain.DefaultIoFilterChain.callPreviousFilterWrite(DefaultIoFilterChain.java:625) ~[mina-core-2.0.9.jar:na]
        at org.apache.mina.core.filterchain.DefaultIoFilterChain.fireFilterWrite(DefaultIoFilterChain.java:618) ~[mina-core-2.0.9.jar:na]
        at org.apache.mina.core.session.AbstractIoSession.write(AbstractIoSession.java:502) ~[mina-core-2.0.9.jar:na]
        at org.apache.mina.core.session.AbstractIoSession.write(AbstractIoSession.java:447) ~[mina-core-2.0.9.jar:na]
        at org.red5.server.net.rtmp.RTMPMinaConnection.write(RTMPMinaConnection.java:328) ~[red5-server-common-1.0.6-SNAPSHOT.jar:1.0.6-SNAPSHOT]
        at org.red5.server.net.rtmp.Channel.write(Channel.java:136) ~[red5-server-common-1.0.6-SNAPSHOT.jar:1.0.6-SNAPSHOT]
        at org.red5.server.net.rtmp.Channel.write(Channel.java:106) ~[red5-server-common-1.0.6-SNAPSHOT.jar:1.0.6-SNAPSHOT]
        at org.red5.server.stream.consumer.ConnectionConsumer.pushMessage(ConnectionConsumer.java:179) ~[red5-server-common-1.0.6-SNAPSHOT.jar:1.0.6-SNAPSHOT]
        at org.red5.server.messaging.InMemoryPushPushPipe.pushMessage(InMemoryPushPushPipe.java:106) ~[red5-server-common-1.0.6-SNAPSHOT.jar:1.0.6-SNAPSHOT]
        at org.red5.server.stream.PlayEngine.doPushMessage(PlayEngine.java:955) ~[red5-server-common-1.0.6-SNAPSHOT.jar:1.0.6-SNAPSHOT]
        at org.red5.server.stream.PlayEngine.sendMessage(PlayEngine.java:1038) ~[red5-server-common-1.0.6-SNAPSHOT.jar:1.0.6-SNAPSHOT]
        at org.red5.server.stream.PlayEngine.pushMessage(PlayEngine.java:1442) ~[red5-server-common-1.0.6-SNAPSHOT.jar:1.0.6-SNAPSHOT]
        at org.red5.server.messaging.InMemoryPushPushPipe.pushMessage(InMemoryPushPushPipe.java:106) ~[red5-server-common-1.0.6-SNAPSHOT.jar:1.0.6-SNAPSHOT]
        at org.bigbluebutton.deskshare.server.ScreenVideoBroadcastStream.dispatchEvent(ScreenVideoBroadcastStream.java:204) ~[ScreenVideoBroadcastStream.class:na]
        at org.bigbluebutton.deskshare.server.stream.DeskshareStream.org$bigbluebutton$deskshare$server$stream$DeskshareStream$$updateStream(DeskshareStream.scala:121) ~[DeskshareStream.class:na]
        at org.bigbluebutton.deskshare.server.stream.DeskshareStream$$anonfun$act$1$$anonfun$apply$1.apply(DeskshareStream.scala:50) ~[DeskshareStream$$anonfun$act$1$$anonfun$apply$1.class:na]
        at org.bigbluebutton.deskshare.server.stream.DeskshareStream$$anonfun$act$1$$anonfun$apply$1.apply(DeskshareStream.scala:47) ~[DeskshareStream$$anonfun$act$1$$anonfun$apply$1.class:na]
        at scala.actors.ReactorTask.run(ReactorTask.scala:31) ~[scala-library-2.9.2.jar:na]
        at scala.actors.Reactor$class.resumeReceiver(Reactor.scala:129) ~[scala-library-2.9.2.jar:na]
```

The latest builds of red5 have a bit better handling of dropping packets.  We're investigating adding that code into the current build of red5.

**Comment by bigbluemanager** [comment]
Just built 1.0beta Saturday from packaging. customer complained of issues with deskshare not working and poor performance

Not sure if this is related to OOM but here is log excerpts

   -- ERRORS found in /usr/share/red5/log/* --
/usr/share/red5/log/error.2016-03-05.log:2016-03-05 16:52:27,064Z [RMI TCP Connection(10)-127.0.0.1] ERROR o.a.c.loader.WebappClassLoaderBase - The web application [deskshare] created a ThreadLocal with key of type [java.lang.ThreadLocal] (value [java.lang.ThreadLocal@76808324]) and a value of type [scala.actors.Actor$$anon$1] (value [scala.actors.Actor$$anon$1@6f21ed43]) but failed to remove it when the web application was stopped. Threads are going to be renewed over time to try and avoid a probable memory leak.

We had 1000’s of these
/usr/share/red5/log/error.2016-03-06.log:2016-03-06 15:20:19,064Z [pool-44-thread-1] ERROR o.r.s.n.r.codec.RTMPProtocolEncoder - Error encoding

 And 
/usr/share/red5/log/red5.2016-03-06.log:java.lang.ClassCastException: null
/usr/share/red5/log/red5.2016-03-06.log:java.lang.ClassCastException: null
/usr/share/red5/log/red5.2016-03-06.log:2016-03-06 21:10:02,545Z [RTMPConnectionExecutor#YDL8TMZXLXOYE-1] WARN  o.r.s.stream.consumer.FileConsumer - Exception waiting for write result. Timeout: 500ms
/usr/share/red5/log/red5.2016-03-06.log:java.lang.InterruptedException: null
/usr/share/red5/log/red5.2016-03-06.log:java.lang.NullPointerException: null

I have not restarted server,  did a quick test meeting and deskshare works for me in test meeting.

**Comment by ffdixon** [comment]
We're not planning to move to the latest red5 in BigBlueButton 1.0.  The version we're using in BigBlueButton 1.0-beta is from August 2015 and we've been testing it under heavy load in production since then.  It's been solid, except for this OOM issue under certain conditions.  

We're now looking at the updates to the latest build of red5 with regards to improvements in dropping of packets and see if we can cherry pick these updates to our version of red5.

This will minimize the changes to the current build and we can test whether they reduce the OOM occurrences in the above scenario.

**Comment by ffdixon** [comment]
(Hmm ... the comment from bigbluemanager has disappeared -- feel free to repost it)

**Comment by ffdixon** [comment]
We believe we found the cause of the error.  In short, if a client receiving a video stream had low bandwidth, the outgoing packets from red5 would back up in the queue and grow unbounded.

This happened with webcam and desktop sharing (both are really just video streams).

After *much* testing and debugging, we found there were two bugs in red5 in how it dropped packets. The 2 lines where the bugs were are in  

https://github.com/bigbluebutton/red5-server-common/blob/7d237a5188f4cdb0e31268ddb449714c7a11263c/src/main/java/org/red5/server/stream/PlayEngine.java#L1497

It was missing the get() where it should have been msgIn.get() but it was only msgIn.getClientBroadcastStream.

The other line is

https://github.com/bigbluebutton/red5-server-common/blob/7d237a5188f4cdb0e31268ddb449714c7a11263c/src/main/java/org/red5/server/stream/PlayEngine.java#L1523

This was always returning a negative value. I changed this to get the size of the mina queue instead.

All other changes are for logging and making the threshold number of packets configurable through spring.

These updates have been merged into red5 master and we've pulled the updated version of red5 with these fixes back into BigBlueButton.

We're *very* happy to mark this issue as closed.
